### ISSUE TITLE
Mathjax preprocessing no longer working in PB 4.x

### ISSUE THREAD

**Issue by monkecheese** [pressbooks/pressbooks#958]
### Prerequisites

_Check the boxes below by changing them from `[ ]` to `[x]`._

* [x] Can you reproduce the behavior you wish to address with [all other plugins deactivated](http://codex.wordpress.org/Multisite_Network_Administration#Plugins)?
* [x] Are you running the [latest version of Pressbooks](http://github.com/pressbooks/pressbooks/releases/latest/) and the [latest version of WordPress](http://codex.wordpress.org/Upgrading_WordPress)?
* [x] Are you experiencing this issue with one of our built-in root or book themes?
* [x] Did you [perform a cursory search](https://github.com/issues?q=+is%3Aissue+repo%3Apressbooks%2Fpressbooks) to see if your bug or enhancement is already listed?

### Description

We are using a plugin that uses the action/filter hooks created in PB 3.9.7 to drop in our own latex renderer (MathJax), [see example](https://github.com/lumenlearning/pressbooks-mathjax).  Now, though MathJax is selectable as a renderer, we were also loading some preprocessing scripts that are no longer working.

### Steps to Reproduce

1. Install/enable mathjax plugin
2. drop in the suggested sample latex on a page: [latex]e^{\i \pi} + 1 = 0[/latex]

**Expected behavior:** The second attached image

**Actual behavior:** The first attached image 

### System Information

Pressbooks: 4.3.2
WordPress: 4.8.2
PHP: 7.0

<img width="135" alt="screen shot 2017-10-03 at 12 23 31 pm" src="https://user-images.githubusercontent.com/10492228/31153615-0af3f6fe-a857-11e7-8d72-010aa9163998.png">
<img width="127" alt="screen shot 2017-10-03 at 12 24 22 pm" src="https://user-images.githubusercontent.com/10492228/31153616-0b0724fe-a857-11e7-9af8-a9c10c9ebce1.png">

**Comment by monkecheese** [comment]
A workaround solution for us is to wrap your `doThisShortcode` in a conditional that checks to see if the `pb_enqueue_latex_scripts` is called:

```
if ( ! has_action( 'pb_enqueue_latex_scripts' ) ) {
	add_filter( 'the_content', array( &$this, 'doThisShortcode' ), 8 ); // Before wpautop()
}
```

I'm not sure if you guys will want this in your code base or not, but I have this solution implemented in our sandbox instance, and would be happy to submit a PR if this solution works for you.  If not, I'm also happy to discuss other possibilities.  Cheers!  :)

**Comment by connerbw** [comment]
Was this working before? If yes, I suspect this PR broke things for you: 

https://github.com/pressbooks/pressbooks/pull/923/files

Can you review what part needs to be put back?

**Comment by connerbw** [comment]
Another thing that changed, we synced the code to WP trunk ([tried to make it look more like this](https://github.com/wp-plugins/wp-latex/commits/master))

Commit is here: https://github.com/pressbooks/pressbooks/commit/763c7de8b7d85cd87622013070f9921f6937ccc3

Of note, this:

```
		add_filter( 'the_content', array( &$this, 'inlineToShortcode' ), 8 );
```

Changed to this:

```
		add_filter( 'the_content', array( &$this, 'inlineToShortcode' ), 7 ); // Before wptexturize()
		add_filter( 'the_content', array( &$this, 'doThisShortcode' ), 8 ); // Before wpautop()
```

Not sure...

**Comment by monkecheese** [comment]
Thanks @connerbw.  I checked out the first code you linked to and tested.  Looks like it was broken at that point.  The second code you linked to seems to be the culprit.  Because when I remove this line, everything fires up and works as expected:

```
add_filter( 'the_content', array( &$this, 'doThisShortcode' ), 8 ); // Before wpautop()
```

**Comment by connerbw** [comment]
@monkecheese 

[Your code](https://github.com/lumenlearning/pressbooks-mathjax/blob/master/latex/mathjax.php), was it [forked from here](https://wordpress.org/plugins/mathjax-latex/#developers)?

Your header says 1.1. Upstream (?) says 1.3.7. They don't look the same.

I don't want to send you on a wild goose chase, but I'm worried about changing this code. It's copy/paste from WordPress. They claim this change was a bugfix to "Fix interactions between LaTeX, `wptexturize()`, and `wpautop()`." ([Source](https://github.com/wp-plugins/wp-latex/commit/54b78b8de0bc05983b3c67785d9f5081f7270d1d))

Is it possible that something in `plugins/mathjax-latex` can be copied over to fix the issue?

**Comment by monkecheese** [comment]
@connerbw 

The problem is in the creation of the shortcode [here](https://github.com/wp-plugins/wp-latex/commit/54b78b8de0bc05983b3c67785d9f5081f7270d1d#diff-c75930f0e8630b8f0f3d35099fa7774bR88).  The config script that we load [here](https://github.com/lumenlearning/pressbooks-mathjax/blob/master/pressbooks-mathjax.php#L74-L87) is looking for `[latex] ... [/latex]` on the page (specifically line 79), and instead it's finding the `<img src='$url' alt='$alt' title='$alt' class='latex' />` being built by the `shortCode` function.

Another possibly better solution might be to add a new filter that allows us to return our own shortcode (bypassing the `shortCode` function) in favor of our own that allows mathjax to render properly.

Example: (pb_add_latex_shortcode)
```
	function doThisShortcode( $text ) {
		$current_shortcodes = $GLOBALS['shortcode_tags'];
		remove_all_shortcodes();

		add_shortcode(
			'latex',
			apply_filters( 'pb_add_latex_shortcode', [ $this, 'shortCode' ] ) );

		$text = do_shortcode( $text );

		$GLOBALS['shortcode_tags'] = $current_shortcodes;

		return $text;
	}
```

**Comment by connerbw** [comment]
What's the simplest solution? Remove the line? I'd rather do that, say with a config variable, than add a new filter that may be deprecated soon (math is on our road map).  Is that OK?

**Comment by connerbw** [comment]
@greatislander 

This? https://github.com/pressbooks/pressbooks/issues/958#issuecomment-334007742

**Comment by monkecheese** [comment]
I think it's simpler for me, right now, to just remove the line as mentioned in my work around solution.  BUT moving forward, if mathjax is something you guys are hoping to add to core, doing the ** pb_add_latex_shortcode** filter hook might be smarter because it's something you might end up using down the line when you circle back on math rendering (per your roadmap).  Either way though, I'm okay with both options.

**Comment by monkecheese** [comment]
Or maybe we can apply the filter at the return inside the `shortCode` function.  Since it's the return format that's really the issue, that might be better still.  Thoughts?

**Comment by monkecheese** [comment]
On second thought, I can appreciate why you wouldn't want to add a new filter that is more or less likely to be deprecated later.  The safer bet might actually be to do the first suggestion for now.

**Comment by greatislander** [comment]
If you wouldn't mind submitting a PR I'd be happy to merge in the AM. Thanks for talking this through with us!

**Comment by monkecheese** [comment]
PR submitted: (#959)[https://github.com/pressbooks/pressbooks/pull/959]

**Comment by monkecheese** [comment]
I just saw the comment about using a config variable instead of checking the pb_enqueue_latex_scripts hook (not sure how I missed that).  If that's preferred, I'd be happy to update the PR. Cheers!

**Comment by greatislander** [comment]
All good, just merged #959.
